### PR TITLE
Introduce `QueryTransform` and add `HyDEQueryTransform`

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* Introduce concept of `QueryProcessor` and simple implementation of HyDE query processor 
* Naive integration with `QueryRunner` where we run query processor once
* Add HyDE example to notebook for demo

### TLDR
```
query_processor = HyDEQueryProcessor(include_original=True)
response = index.query("what did author do after going to RISD", query_processor=query_processor)
``` 

### Future work
- Support generating multiple hypothetical docs
- Consider refactor to support index-specific query processing

### CODE REVIEW

File: gpt_index/indices/query/query_runner.py [hunk 1]
```diff
@@ -74,9 +77,10 @@ def query(
         self, query_str_or_bundle: Union[str, QueryBundle], index_struct: IndexStruct
     ) -> Response:
         """Run query."""
+        # NOTE: query processor is only run once
+        # TODO: Consider refactor to support index-specific query processing
```
Comment (jerryjliu) [inline gpt_index/indices/query/query_runner.py:81]: was going to ask this, but this makes sense. i'm potentially interested in how we can do query processing given contents of the nodes themselves

File: examples/vector_indices/SimpleIndexDemo.ipynb [hunk 2]
```diff
@@ -151,6 +151,73 @@
     "display(Markdown(f\"<b>{response}</b>\"))"
    ]
   },
+  {
```
Comment (jerryjliu) [inline examples/vector_indices/SimpleIndexDemo.ipynb:154]: The notebook organization is a mess atm, but imo I'd create a new folder under `examples` called QueryTransformations which specifically demos HyDE.

File: gpt_index/indices/query/query_processor.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    A query processor augments a raw query string with associated transformations
+    to improve index querying.
+    """
+
+    def __call__(self, query_str: str) -> QueryBundle:
+        """Run query processor."""
+        return QueryBundle(query_str=query_str, embedding_strs=[query_str])
+
+
+class HyDEQueryProcessor(BaseQueryProcessor):
```
Comment (jerryjliu) [inline gpt_index/indices/query/query_processor.py:23]: would love to have a How-To Guide + Reference section in the docs. 

reference could go in docs/reference/query.rst

**Review by jerryjliu — COMMENTED** [review]
this looks pretty good! i'll approve after the first PR lands

File: gpt_index/indices/query/query_runner.py (hunk 1 above)
Comment (Disiok) [reply gpt_index/indices/query/query_runner.py]: Yep, going to keep this PR small. Definitely interesting future work to improve query processing with node content (and or metadata).

File: examples/vector_indices/SimpleIndexDemo.ipynb (hunk 2 above)
Comment (Disiok) [reply examples/vector_indices/SimpleIndexDemo.ipynb]: Added a separate notebook with HyDE examples.

File: gpt_index/indices/query/query_processor.py (hunk 3 above)
Comment (Disiok) [reply gpt_index/indices/query/query_processor.py]: Updated reference.

**Comment by Disiok** [comment]
@jerryjliu  See the notebook for 3 examples comparing normal vs. HyDE.

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**